controller/templates/controller/partials/state_machine.html

```diff
@@ -1,6 +1,42 @@
 {% load render_table from django_tables2 %}
 <div _="on load if #argsDialog is not null remove #argsDialog"></div>
-<form>
-  {% csrf_token %}
-  {% render_table table %}
-</form>
+<div class="row">
+  <div class="col-md-auto">
+    <form>
+      {% csrf_token %}
+      {% render_table table %}
+    </form>
+  </div>
+  <div class="col-md-auto">
+    <pre class="mermaid">
+      flowchart TD
+
+      %% FSM states (nodes).
+      classDef default fill:#b5b3ae,stroke:black,stroke-width:2px
+
+      %% FSM transitions (edges).
+      linkStyle default background-color:#b5b3ae,stroke-width:2px
+
+      {{ flowchart }}
+    </pre>
+  </div>
+</div>
+<script>
+  // Initialize mermaid diagrams.
+  mermaid.initialize({
+    theme: "base",
+    // Make flowchart text uppercase.
+    themeCSS: ".label { text-transform: uppercase; }",
+    themeVariables: {
+      primaryTextColor: "white",
+      edgeLabelBackground: "transparent", // HACK: custom edge label colour
+    },
+  });
+  // Refresh mermaid diagrams after htmx swap.
+  document.body.addEventListener("htmx:afterSwap", (event) => {
+      const target = event.detail.target;
+      if (target && target.querySelector(".mermaid")) {
+          mermaid.run({nodes: target.querySelectorAll(".mermaid")});
+      }
+  });
```

Comment:
I don't get this. Why does the diagram need to be manually refreshed? The whole block should be re-loaded and, presumably, a brand new diagram created with the new content.

Reply:
The problem with mermaid is the diagrams are only rendered once, when the *main* document finishes loading. Content swaps with HTMX posts do not trigger another rendering, so it has to be this way.